src/main/java/edu/umich/its/cpm/Utils.java

```diff
@@ -443,19 +443,46 @@ public static String getUserLoginId(HttpServletRequest request, Environment env)
 		boolean isTestUrlEnabled = Boolean.valueOf(env.getProperty(ALLOW_USER_URLOVERRIDE));
 		String testUserInSession = (String) request.getSession().getAttribute(TEST_USER);
 
-		if (isTestUrlEnabled && testUser != null) {
+		//become user case
+		if (isTestUrlEnabled && testUser != null && remoteUser != null && testUser != remoteUser) {
+			if (isCurrentUserCPMAdmin(remoteUser, env)) {
+				request.getSession().setAttribute(TEST_USER, testUser);
+				user = testUser;
+				log.debug("BECOME USER " + user);
+			} else {
+				user = remoteUser;
+				log.error("the user " + user + " is unauthorized to become user");
+			}
+			return user;
+		}
+		//become user case
+		if (isTestUrlEnabled && testUserInSession != null && remoteUser != null && testUserInSession != remoteUser) {
+			if (isCurrentUserCPMAdmin(remoteUser, env)) {
+				user = testUserInSession;
```

Comment:
Should this set the TEST_USER attribute?  If no put in a comment about why this isn't symmetric with the prior if.